crates/sui-graphql-rpc/src/types/move_module.rs

```diff
@@ -46,12 +48,111 @@ impl MoveModule {
         }
     }
 
-    // friendConnection(
-    //   first: Int,
-    //   after: String,
-    //   last: Int,
-    //   before: String
-    // ): MoveModuleConnection
+    /// Modules that this module considers friends (these modules can access `public(friend)`
+    /// functions from this module).
+    async fn friend_connection(
+        &self,
+        ctx: &Context<'_>,
+        first: Option<u64>,
+        after: Option<String>,
+        last: Option<u64>,
+        before: Option<String>,
+    ) -> Result<Connection<String, MoveModule>> {
+        // TODO: make cursor opaque (currently just an offset).
+        validate_cursor_pagination(&first, &after, &last, &before).extend()?;
+
+        let bytecode = self.parsed.bytecode();
+        let total = bytecode.friend_decls.len();
+
+        // Add one to make [lo, hi) a half-open interval ((after, before) is an open interval).
+        let mut lo = if let Some(after) = after {
+            1 + after
+                .parse::<usize>()
+                .map_err(|_| Error::InvalidCursor("Failed to parse 'after' cursor.".to_string()))
+                .extend()?
+        } else {
+            0
+        };
+
+        let mut hi = if let Some(before) = before {
+            before
+                .parse::<usize>()
+                .map_err(|_| Error::InvalidCursor("Failed to parse 'before' cursor.".to_string()))
+                .extend()?
+        } else {
+            total
+        };
+
+        let mut connection = Connection::new(false, false);
+        if hi <= lo {
+            return Ok(connection);
+        }
+
+        if let Some(first) = first {
+            let first = first as usize;
+            if hi - lo > first {
+                hi = lo + first;
```

Comment:
took me a while to process that we're just bounding the upper bound by lo + first